### PR TITLE
Relation id and afterAll hook performance fixes

### PR DESCRIPTION
**Pull request by Lennard-Dietz**
This PR aims to increase the performance of some very specific typeorm parts.
This includes the entity transformation after a query, when using the `@RelationId` decorator as well as broadcasting the `afterAll` event to all subscribers.
All it does is refactor some internal class logic for performance optimization. Therefore no new tests are needed, as the results of all the processes are still the same.

**Motivation**
When using typeorm, we saw that running queries which return > 500 entities are having a very bad performance and in our case, as we are using the `@RelationId` decorator a lot, we even saw a quadratic increase in runtime of the queries when increasing the number of entities.
Therefore we did a deep dive and found several performance optimizations. By far the biggest one is in the `RawSqlResultsToEntityTransformer`, as there was a quadratic calculation when using `@RelationId`. We were able to get this down to a linear runtime, which helped increase the performance of our queries a lot, around 3 times faster when returning 1000 entities.

It is related to the following issues, depending on the used entities, but does not fix them completely:
#2381
#3857

tl;dr

- Replace quadratic runtime in `RawSqlResultsToEntityTransformer` with linear runtime
- Use object instead of array for faster duplicate checks in `RelationIdLoader`
- Calculate the fitting subscribers for the given entity only once and not for every result in `Broadcaster`

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change "N/A"
- [ ] Documentation has been updated to reflect this change "N/A"
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by antoniusostermann** [comment]
This would solve a lot of our problems! Any plans on when to merge this?

File: src/query-builder/relation-id/RelationIdLoader.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export class RelationIdLoader {
                 if (relationIdAttr.queryBuilderFactory)
                     throw new TypeORMError("Additional condition can not be used with ManyToOne or OneToOne owner relations.");
 
-                const duplicates: Array<string> = [];
+                const duplicates: {[duplicateKey: string]: boolean} = {};
```
[the remark this answers is not in the document]
Comment (Lennard-Dietz) [reply src/query-builder/relation-id/RelationIdLoader.ts:36]: A set would perform much worse in this case

File: src/query-builder/relation-id/RelationIdLoader.ts [hunk 2]
```diff
@@ -78,7 +78,7 @@ export class RelationIdLoader {
                 const tableName = relation.inverseEntityMetadata.tableName; // category
                 const tableAlias = relationIdAttr.alias || tableName; // if condition (custom query builder factory) is set then relationIdAttr.alias defined
 
-                const duplicates: Array<string> = [];
+                const duplicates: {[duplicateKey: string]: boolean} = {};
```
[the remark this answers is not in the document]
Comment (Lennard-Dietz) [reply src/query-builder/relation-id/RelationIdLoader.ts:81]: See [comment](https://github.com/typeorm/typeorm/pull/8169/files#r756029041)

File: src/query-builder/relation-id/RelationIdLoader.ts [hunk 3]
```diff
@@ -174,7 +174,7 @@ export class RelationIdLoader {
                     return { relationIdAttribute: relationIdAttr, results: [] };
 
                 const parameters: ObjectLiteral = {};
-                const duplicates: Array<string> = [];
+                const duplicates: {[duplicateKey: string]: boolean} = {};
```
[the remark this answers is not in the document]
Comment (Lennard-Dietz) [reply src/query-builder/relation-id/RelationIdLoader.ts:177]: See [comment](https://github.com/typeorm/typeorm/pull/8169/files#r756029041)

File: src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts [hunk 4]
```diff
@@ -378,4 +349,82 @@ export class RawSqlResultsToEntityTransformer {
...
+                // Calculate column metadata
+                let columns: ColumnMetadata[];
+                if (relation.isManyToOne || relation.isOneToOneOwner) {
+                    columns = relation.joinColumns.map(joinColumn => joinColumn);
+                } else if (relation.isOneToMany || relation.isOneToOneNotOwner) {
+                    columns = relation.inverseEntityMetadata.primaryColumns.map(joinColumn => joinColumn);
+                } else {
+                    // ManyToMany
+                    if (relation.isOwning) {
+                    columns = relation.inverseJoinColumns.map(joinColumn => joinColumn);
```
[the remark this answers is not in the document]
Comment (Lennard-Dietz) [reply src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts]: This was just a copy of the previous data transformations, but you're right, these maps can be removed

File: src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts [hunk 5]
```diff
@@ -378,4 +349,82 @@ export class RawSqlResultsToEntityTransformer {
...
+                    } else {
+                    columns = relation.inverseRelation!.joinColumns.map(joinColumn => joinColumn);
```
[the remark this answers is not in the document]
Comment (Lennard-Dietz) [reply src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts]: See [comment](https://github.com/typeorm/typeorm/pull/8169/files#r756079648)

**Comment by Lennard-Dietz** [comment]
Thank you @rluvaton for your little review 👍

**Review by Pho3niX90 — APPROVED** [review]

**Review by Pho3niX90 — APPROVED** [review]
